executor/aggregate.go

```diff
@@ -45,39 +96,484 @@ type HashAggExec struct {
+	}
+
+	// Init final workers.
+	finalAggFuncs := e.newFinalAggFuncs()
+	for i := 0; i < finalConcurrency; i++ {
+		baseAggWorker := baseHashAggWorker{
+			finishCh:     e.finishCh,
+			aggFuncs:     finalAggFuncs,
+			maxChunkSize: e.maxChunkSize,
+			groupVals:    make([][]byte, 0, 8),
+		}
+		e.finalWorkers[i] = HashAggFinalWorker{
+			baseHashAggWorker:   baseAggWorker,
+			aggCtxsMap:          make(aggCtxsMapper, 0),
+			groupMap:            mvmap.NewMVMap(),
+			inputCh:             e.partialOutputChs[i],
+			outputCh:            e.finalOutputCh,
+			finalResultHolderCh: make(chan *chunk.Chunk, 1),
+			rowBuffer:           make([]types.Datum, 0, e.Schema().Len()),
+			mutableRow:          chunk.MutRowFromTypes(e.retTypes()),
+		}
+		e.finalWorkers[i].finalResultHolderCh <- e.newChunk()
+	}
+}
+
+func (e *HashAggExec) newFinalAggFuncs() (newAggFuncs []aggregation.Aggregation) {
+	newAggFuncs = make([]aggregation.Aggregation, 0, len(e.AggFuncs))
+	idx := 0
+	for _, af := range e.AggFuncs {
+		var aggFunc aggregation.Aggregation
+		idx, aggFunc = af.GetFinalAggFunc(idx)
+		newAggFuncs = append(newAggFuncs, aggFunc)
+	}
+	return newAggFuncs
+}
+
+// HashAggPartialWorker gets and handles origin data or partial data from inputCh,
+// then shuffle the intermediate results to corresponded final workers.
+func (w *HashAggPartialWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup) {
+	var (
+		chk         *chunk.Chunk
+		ok          bool
+		sc          = ctx.GetSessionVars().StmtCtx
+		needShuffle = false
+	)
+	defer func() {
+		if needShuffle {
+			w.shuffleIntermData(sc)
+		}
+		waitGroup.Done()
+	}()
+	for {
+		select {
+		case <-w.finishCh:
+			return
+		case chk, ok = <-w.inputCh:
+			if !ok {
+				return
+			}
+		}
+
+		if !w.updateIntermData(ctx, sc, chk, len(w.aggCtxsMaps)) {
+			return
+		}
+		needShuffle = true
+		if chk.NumRows() == 0 {
+			return
+		}
+		w.giveBackCh <- &HashAggInput{
+			chk:        chk,
+			giveBackCh: w.inputCh,
+		}
+	}
+}
+
+func (w *HashAggPartialWorker) updateIntermData(ctx sessionctx.Context, sc *stmtctx.StatementContext, chk *chunk.Chunk, finalConcurrency int) (isOk bool) {
+	inputIter := chunk.NewIterator4Chunk(chk)
+	for row := inputIter.Begin(); row != inputIter.End(); row = inputIter.Next() {
+		groupKey, err := w.getGroupKey(sc, row)
+		if err != nil {
+			w.globalOutputCh <- &AfFinalResult{err: errors.Trace(err)}
+			return false
+		}
+		finalWorkerIdx := int(murmur3.Sum32(groupKey)) % finalConcurrency
+		if w.aggCtxsMaps[finalWorkerIdx] == nil {
+			w.aggCtxsMaps[finalWorkerIdx] = make(aggCtxsMapper, 0)
+			w.groupSets[finalWorkerIdx] = mvmap.NewMVMap()
+		}
+		aggCtxMap, groupSet := w.aggCtxsMaps[finalWorkerIdx], w.groupSets[finalWorkerIdx]
+		if len(groupSet.Get(groupKey, w.groupVals[:0])) == 0 {
+			groupSet.Put(groupKey, []byte{})
+		}
+		aggEvalCtxs := w.getContext(sc, groupKey, aggCtxMap)
+		for i, af := range w.aggFuncs {
+			err = af.Update(aggEvalCtxs[i], sc, row)
+			if err != nil {
```

Comment:
If `err != nil`, I think we can just return `errors.Trace(err)`, and handle the error outside, which can help us to reduce the duplicated code to handle error in this function.